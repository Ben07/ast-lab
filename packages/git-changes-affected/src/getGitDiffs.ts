import exec from './exec';
import { GIT_OPERATION } from './constants';
import { Diff } from './types';
import getRevisionFile from './getRevisionFile';

function justifyOperation(operatelog: string) {
  if (/^new file mode/.test(operatelog)) {
    return GIT_OPERATION.new;
  }
  if (/^delete file mode/.test(operatelog)) {
    return GIT_OPERATION.delete;
  }
  /** @todo confirm rename similarity percentage */
  if (/^similarity index/.test(operatelog)) {
    return GIT_OPERATION.rename;
  }
  return GIT_OPERATION.change;
}
function isLineRemoved(content: string) {
  return content.startsWith('-');
}
function isLineAdded(content: string) {
  return content.startsWith('+');
}

/**
 * @todo get real diff content
 * @param {string} commit 
 */
export function getGitDiffs(commit: string): Diff[] {
  const strOut = exec(`git show ${commit} --format="%N" --first-parent`);
  const diffs = [] as Diff[];
  let lineA = 0;
  let lineB = 0;
  let aChangeStart = null as number | null;
  let bChangeStart = null as number | null;
  const diffLines = strOut.split('\n')
  diffLines.forEach((content, index) => {
    const lastDiff = diffs[diffs.length - 1];
    const aChanges = lastDiff &&lastDiff.source.changed;
    const bChanges = lastDiff &&lastDiff.target.changed;
    const fileHeadMatch = content.match(/^diff --git a\/([^\n\s]+) b\/([^\n\s]+)/);
    if (fileHeadMatch) {
      if (lastDiff) {
        if (
          lastDiff.operation !== GIT_OPERATION.rename && 
          aChangeStart !== null
        ) {
          aChanges.push({ 
            start: aChangeStart,
            end: lineA - 1
          });
        }
        if (bChangeStart !== null) {
          bChanges.push({
            start: bChangeStart, 
            end: lineB - 1
          });
        }
      }
      const [sourceFile, targetFile] = fileHeadMatch.slice(1);
      const operation = justifyOperation(diffLines[index + 1]);
      diffs.push({
        source: {
          file: sourceFile,
          content: operation === GIT_OPERATION.new ? null : getRevisionFile(`${commit}~1`, sourceFile),
          changed: operation === GIT_OPERATION.rename ? [{
            start: 0, 
            end: Infinity
          }]: []
        },
        target: {
          file: targetFile,
          content:  operation === GIT_OPERATION.delete ? null : getRevisionFile(commit, targetFile),
          changed: []
        },
        operation,
      });
      return;
    }
    const chunkHeadMatch = content.match(/@@ -(\d+),\d+ \+(\d+),\d+ @@/);
    if (chunkHeadMatch) {
      lineA = parseInt(chunkHeadMatch[1], 10);
      lineB = parseInt(chunkHeadMatch[2], 10);
      return;
    }
    const isLastLine = index === diffLines.length - 1;
    const isAdded = isLineAdded(content);
    const isRemoved = isLineRemoved(content);
    if (
      lastDiff.operation !== GIT_OPERATION.rename
    ) {
      if (isRemoved && aChangeStart === null) {
        aChangeStart = lineA;
      }
      if (
        (!isRemoved || isLastLine) && aChangeStart !== null
      ) {
        aChanges.push({ 
          start: aChangeStart,
          end: lineA - 1
        });
        aChangeStart = null;
      }
      if (!isAdded) {
        lineA++;
      }
    }
    if (isAdded && bChangeStart === null) {
      bChangeStart = lineB;
    }
    if (
      (!isAdded || isLastLine) && 
      bChangeStart !== null
    ) {
      bChanges.push({
        start: bChangeStart, 
        end: lineB - 1
      });
      bChangeStart = null;
    }
    if (!isRemoved) {
      lineB++;
    }
  });
  return diffs;
}

export default getGitDiffs;