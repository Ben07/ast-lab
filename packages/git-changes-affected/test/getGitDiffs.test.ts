import getGitDiffs from '../src/getGitDiffs';

describe('getGitDiffs', () => {
    test('e42f4f6a08569afd48540759cc604aab2b8c02f3', () => {
      expect(getGitDiffs('e42f4f6a08569afd48540759cc604aab2b8c02f3')).toMatchSnapshot();
    });
    test('7037cc7c2229d424df63cdbd0981463d1801c697', () => {
      expect(getGitDiffs('7037cc7c2229d424df63cdbd0981463d1801c697')).toMatchSnapshot();
    });
});
