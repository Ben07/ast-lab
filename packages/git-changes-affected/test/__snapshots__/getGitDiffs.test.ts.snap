// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGitDiffs 7037cc7c2229d424df63cdbd0981463d1801c697 1`] = `
Array [
  Object {
    "operation": 3,
    "source": Object {
      "changed": Array [
        Object {
          "end": Infinity,
          "start": 0,
        },
      ],
      "content": "import { DependencyMap, Entry } from \\"ast-lab-types\\";

type Visited = { [module: string]: { [member: string]: 1 } };

export default function visitDependencyMap(dependencyMap: DependencyMap, entries: Entry[]): Visited {
  const visited = {} as Visited;
  let entryQueue = entries;
  while(entryQueue.length) {
    const { source: mod, name } = entryQueue.shift() as Entry;
    let modVisited = visited[mod];
    if (!modVisited) {
      modVisited = visited[mod] = {};
    }
    const affected = dependencyMap.get(mod);
    if (affected && modVisited[name]) {
      const affectedEntries = affected.get(name);
      if (affectedEntries && affectedEntries.length) {
        entryQueue = entryQueue.concat(affectedEntries);
      }
    }
  }
  return visited;
}",
      "file": "packages/get-dependencies/src/visitDependencyMap.ts",
    },
    "target": Object {
      "changed": Array [
        Object {
          "end": 4,
          "start": 2,
        },
        Object {
          "end": 11,
          "start": 11,
        },
        Object {
          "end": 21,
          "start": 20,
        },
        Object {
          "end": 24,
          "start": 23,
        },
      ],
      "content": "import { DependencyMap, Entry } from \\"ast-lab-types\\";
import _debug from 'debug';

const debug = _debug('get-dependencies:visit');

type Visited = { [module: string]: { [member: string]: 1 } };

export default function visitDependencyMap(dependencyMap: DependencyMap, entries: Entry[]): Visited {
  const visited = {} as Visited;
  let entryQueue = entries;
  debug('visit entries:', entries);
  while(entryQueue.length) {
    const { source: mod, name } = entryQueue.shift() as Entry;
    let modVisited = visited[mod];
    if (!modVisited) {
      modVisited = visited[mod] = {};
    }
    const affected = dependencyMap.get(mod);
    debug(\`\${mod} affects\`, affected);
    if (affected && !modVisited[name]) {
      const affectedEntries = affected.get(name);
      debug('current queue:', entryQueue);
      debug('new to queue:', affectedEntries);
      if (affectedEntries && affectedEntries.length) {
        entryQueue = entryQueue.concat(affectedEntries);
      }
    }
  }
  return visited;
}",
      "file": "packages/get-dependencies/src/visitDepMap.ts",
    },
  },
  Object {
    "operation": 1,
    "source": Object {
      "changed": Array [],
      "content": null,
      "file": "packages/get-dependencies/test/__snapshots__/visitDepMap.test.ts.snap",
    },
    "target": Object {
      "changed": Array [
        Object {
          "end": 21,
          "start": 1,
        },
      ],
      "content": "// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[\`visitDepMap() circular 1\`] = \`
Object {
  \\"/a\\": Object {},
}
\`;

exports[\`visitDepMap() importDefault.* 1\`] = \`
Object {
  \\"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAll.js\\": Object {},
  \\"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importDefault.js\\": Object {},
}
\`;

exports[\`visitDepMap() importNamed.resolveFactory 1\`] = \`
Object {
  \\"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAlias.js\\": Object {},
  \\"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importNamed.js\\": Object {},
}
\`;",
      "file": "packages/get-dependencies/test/__snapshots__/visitDepMap.test.ts.snap",
    },
  },
  Object {
    "operation": 1,
    "source": Object {
      "changed": Array [],
      "content": null,
      "file": "packages/get-dependencies/test/visitDepMap.test.ts",
    },
    "target": Object {
      "changed": Array [
        Object {
          "end": 102,
          "start": 1,
        },
      ],
      "content": "import visitDepMap from '../src/visitDepMap';

describe('visitDepMap()', () => {
  const depMap = new Map([
    [
      \\"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAll.js\\",
      new Map([[\\"b\\", []]])
    ], 
    [
    \\"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importDefault.js\\",
    new Map([
      [
        \\"*\\", 
        [{
          \\"name\\": \\"a\\",
          \\"source\\": \\"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAll.js\\",
        }, {
          \\"name\\": \\"func\\",
          \\"source\\": \\"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAll.js\\",
        }],
      ],
    ]),
  ],
  [
    \\"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importNamed.js\\",
    new Map([
      [
        \\"resolveFactory\\", 
        [{
          \\"name\\": \\"c\\",
          \\"source\\": \\"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAlias.js\\",
        },
        {
          \\"name\\": \\"func\\",
          \\"source\\": \\"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAlias.js\\",
        }],
      ]
    ])
  ]
  ]);
  test('importNamed.resolveFactory', () => {
    expect(visitDepMap(depMap, [
      {
        name: 'resolveFactory',
        source: '/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importNamed.js'
      }
    ])).toMatchSnapshot();
  });
  test('importDefault.*', () => {
    expect(visitDepMap(depMap, [
      {
        name: '*',
        source: '/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importDefault.js'
      }
    ])).toMatchSnapshot();
  });
  
  // test('circular', () => {
  //   expect(visitDepMap(
  //     new Map([
  //       [
  //         '/a',
  //         new Map([
  //           [
  //             'default',
  //             [{
  //               source: '/b',
  //               name: 'default'
  //             }]
  //           ]
  //         ])
  //       ], [
  //         '/b',
  //         new Map([
  //           [
  //             'default',
  //             [{
  //               source: '/c',
  //               name: 'default'
  //             }]
  //           ]
  //         ])
  //       ], [
  //         '/c',
  //         new Map([
  //           [
  //             'default',
  //             [{
  //               source: '/a',
  //               name: 'default'
  //             }]
  //           ]
  //         ])
  //       ]
  //     ]), 
  //     [{
  //       name: 'default',
  //       source: '/a'
  //     }]
  //   )).toMatchSnapshot();
  // });
});",
      "file": "packages/get-dependencies/test/visitDepMap.test.ts",
    },
  },
]
`;

exports[`getGitDiffs e42f4f6a08569afd48540759cc604aab2b8c02f3 1`] = `
Array [
  Object {
    "operation": 0,
    "source": Object {
      "changed": Array [],
      "content": "export { default } from './getStats';",
      "file": "packages/es-stats/src/index.ts",
    },
    "target": Object {
      "changed": Array [],
      "content": "export { default as createExportVisitors } from './visitors/exports';
export { default as createImportVisitors } from './visitors/imports';
export { default as createRootRelationVisitors } from './visitors/rootRelation';
export { default as  mergeVisitors } from './mergeVisitors';
export { default } from './getStats';",
      "file": "packages/es-stats/src/index.ts",
    },
  },
  Object {
    "operation": 0,
    "source": Object {
      "changed": Array [
        Object {
          "end": 14,
          "start": 14,
        },
        Object {
          "end": 20,
          "start": 20,
        },
        Object {
          "end": 29,
          "start": 29,
        },
        Object {
          "end": 35,
          "start": 35,
        },
        Object {
          "end": 42,
          "start": 42,
        },
        Object {
          "end": 45,
          "start": 45,
        },
      ],
      "content": "import { Visitor } from '@babel/traverse';
import getModuleRefFromExportSpecifier from '../getModuleRefFromExportSpecifier';
import getDeclarationNames from '../getDeclarationNames';
import { MODULE_DEFAULT } from '../constants';
import { Exports } from 'ast-lab-types';

export default function createExportVisitors(exports: Exports = { members: [] }): Visitor {
  return {
    ExportAllDeclaration({ node }) {
      exports.extends = (exports.extends || []).concat(node.source.value);
    },
    ExportNamedDeclaration({ node }) {
      const { specifiers, declaration } = node;
      if (specifiers.length) {
        specifiers.forEach(specifier => {
          // @ts-ignore
          const dep = getModuleRefFromExportSpecifier(specifier);
          if (dep) {
            exports.members.push(dep);
          }
        });
      }
      if (declaration) {
        // @ts-ignore
        const names = getDeclarationNames(declaration)
        if (names && names.length) {
          names.forEach(({ name }) => {
            exports.members.push({ name, alias: name });
          });
        }
      }
    },
    ExportDefaultDeclaration({ node }) {
      const { declaration } = node;
      const alias = MODULE_DEFAULT;
      // @ts-ignore
      const names = getDeclarationNames(declaration);
      if (names && names.length) {
        names.forEach(({ name }) => {
          name = name || MODULE_DEFAULT;
          exports.members.push({ name, alias });
        });
      } else {
        exports.members.push({ name: MODULE_DEFAULT, alias: MODULE_DEFAULT });
      }
    }
  };
}",
      "file": "packages/es-stats/src/visitors/exports.ts",
    },
    "target": Object {
      "changed": Array [
        Object {
          "end": 14,
          "start": 14,
        },
        Object {
          "end": 23,
          "start": 20,
        },
        Object {
          "end": 32,
          "start": 32,
        },
        Object {
          "end": 38,
          "start": 38,
        },
        Object {
          "end": 45,
          "start": 45,
        },
        Object {
          "end": 48,
          "start": 48,
        },
      ],
      "content": "import { Visitor } from '@babel/traverse';
import getModuleRefFromExportSpecifier from '../getModuleRefFromExportSpecifier';
import getDeclarationNames from '../getDeclarationNames';
import { MODULE_DEFAULT } from '../constants';
import { Exports } from 'ast-lab-types';

export default function createExportVisitors(exports: Exports = { members: [] }): Visitor {
  return {
    ExportAllDeclaration({ node }) {
      exports.extends = (exports.extends || []).concat(node.source.value);
    },
    ExportNamedDeclaration({ node }) {
      const { specifiers, declaration, loc } = node;
      if (specifiers.length) {
        specifiers.forEach(specifier => {
          // @ts-ignore
          const dep = getModuleRefFromExportSpecifier(specifier);
          if (dep) {
            exports.members.push({
              ...dep,
              loc
            });
          }
        });
      }
      if (declaration) {
        // @ts-ignore
        const names = getDeclarationNames(declaration)
        if (names && names.length) {
          names.forEach(({ name }) => {
            exports.members.push({ name, alias: name, loc });
          });
        }
      }
    },
    ExportDefaultDeclaration({ node }) {
      const { declaration, loc } = node;
      const alias = MODULE_DEFAULT;
      // @ts-ignore
      const names = getDeclarationNames(declaration);
      if (names && names.length) {
        names.forEach(({ name }) => {
          name = name || MODULE_DEFAULT;
          exports.members.push({ name, alias, loc });
        });
      } else {
        exports.members.push({ name: MODULE_DEFAULT, alias: MODULE_DEFAULT,loc });
      }
    }
  };
}",
      "file": "packages/es-stats/src/visitors/exports.ts",
    },
  },
  Object {
    "operation": 0,
    "source": Object {
      "changed": Array [],
      "content": "// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[\`export visitors exportAllFrom.js 1\`] = \`
Object {
  \\"extends\\": Array [
    \\"./exportNamed\\",
  ],
  \\"members\\": Array [],
}
\`;

exports[\`export visitors exportDefault.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"default\\",
      \\"name\\": \\"default\\",
    },
  ],
}
\`;

exports[\`export visitors exportDefaultFuncAlias.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"default\\",
      \\"name\\": \\"default\\",
    },
  ],
}
\`;

exports[\`export visitors exportDefaultValueAlias.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"default\\",
      \\"name\\": \\"default\\",
    },
  ],
}
\`;

exports[\`export visitors exportFunc.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"a\\",
      \\"name\\": \\"a\\",
    },
  ],
}
\`;

exports[\`export visitors exportFuncAlias.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"b\\",
      \\"name\\": \\"a\\",
    },
  ],
}
\`;

exports[\`export visitors exportNamed.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"a\\",
      \\"name\\": \\"a\\",
    },
  ],
}
\`;

exports[\`export visitors exportNamedFrom.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"a\\",
      \\"name\\": \\"a\\",
    },
  ],
}
\`;

exports[\`export visitors noExports.js 1\`] = \`
Object {
  \\"members\\": Array [],
}
\`;",
      "file": "packages/es-stats/test/__snapshots__/visitor.exports.test.ts.snap",
    },
    "target": Object {
      "changed": Array [
        Object {
          "end": 27,
          "start": 18,
        },
        Object {
          "end": 48,
          "start": 39,
        },
        Object {
          "end": 69,
          "start": 60,
        },
        Object {
          "end": 90,
          "start": 81,
        },
        Object {
          "end": 111,
          "start": 102,
        },
        Object {
          "end": 132,
          "start": 123,
        },
        Object {
          "end": 153,
          "start": 144,
        },
      ],
      "content": "// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[\`export visitors exportAllFrom.js 1\`] = \`
Object {
  \\"extends\\": Array [
    \\"./exportNamed\\",
  ],
  \\"members\\": Array [],
}
\`;

exports[\`export visitors exportDefault.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"default\\",
      \\"loc\\": SourceLocation {
        \\"end\\": Position {
          \\"column\\": 17,
          \\"line\\": 3,
        },
        \\"start\\": Position {
          \\"column\\": 0,
          \\"line\\": 3,
        },
      },
      \\"name\\": \\"default\\",
    },
  ],
}
\`;

exports[\`export visitors exportDefaultFuncAlias.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"default\\",
      \\"loc\\": SourceLocation {
        \\"end\\": Position {
          \\"column\\": 17,
          \\"line\\": 5,
        },
        \\"start\\": Position {
          \\"column\\": 0,
          \\"line\\": 5,
        },
      },
      \\"name\\": \\"default\\",
    },
  ],
}
\`;

exports[\`export visitors exportDefaultValueAlias.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"default\\",
      \\"loc\\": SourceLocation {
        \\"end\\": Position {
          \\"column\\": 17,
          \\"line\\": 3,
        },
        \\"start\\": Position {
          \\"column\\": 0,
          \\"line\\": 3,
        },
      },
      \\"name\\": \\"default\\",
    },
  ],
}
\`;

exports[\`export visitors exportFunc.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"a\\",
      \\"loc\\": SourceLocation {
        \\"end\\": Position {
          \\"column\\": 1,
          \\"line\\": 3,
        },
        \\"start\\": Position {
          \\"column\\": 0,
          \\"line\\": 1,
        },
      },
      \\"name\\": \\"a\\",
    },
  ],
}
\`;

exports[\`export visitors exportFuncAlias.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"b\\",
      \\"loc\\": SourceLocation {
        \\"end\\": Position {
          \\"column\\": 16,
          \\"line\\": 5,
        },
        \\"start\\": Position {
          \\"column\\": 0,
          \\"line\\": 5,
        },
      },
      \\"name\\": \\"a\\",
    },
  ],
}
\`;

exports[\`export visitors exportNamed.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"a\\",
      \\"loc\\": SourceLocation {
        \\"end\\": Position {
          \\"column\\": 19,
          \\"line\\": 1,
        },
        \\"start\\": Position {
          \\"column\\": 0,
          \\"line\\": 1,
        },
      },
      \\"name\\": \\"a\\",
    },
  ],
}
\`;

exports[\`export visitors exportNamedFrom.js 1\`] = \`
Object {
  \\"members\\": Array [
    Object {
      \\"alias\\": \\"a\\",
      \\"loc\\": SourceLocation {
        \\"end\\": Position {
          \\"column\\": 34,
          \\"line\\": 1,
        },
        \\"start\\": Position {
          \\"column\\": 0,
          \\"line\\": 1,
        },
      },
      \\"name\\": \\"a\\",
    },
  ],
}
\`;

exports[\`export visitors noExports.js 1\`] = \`
Object {
  \\"members\\": Array [],
}
\`;",
      "file": "packages/es-stats/test/__snapshots__/visitor.exports.test.ts.snap",
    },
  },
  Object {
    "operation": 0,
    "source": Object {
      "changed": Array [
        Object {
          "end": 11,
          "start": 11,
        },
        Object {
          "end": 14,
          "start": 14,
        },
      ],
      "content": "import { ParserOptions } from '@babel/parser';

export type Module = string;
export type Member = string;
export type Members = Member[] | '*';
export type MemberRef = {
  name: Member,
  alias: Member,
};
export type Import =  MemberRef & Entry;
export type Exports = {
  extends?: Module[],
  members: MemberRef[]
};
export type MemberRelation =  { [name: string]: Member[] };

export type Entry = {
  source: Module, 
  name: Member
};
export type AffectedMap = Map<Member, Entry[]>;
export type DependencyMap = Map<Module, AffectedMap>;

export type Loader = (path: string) => Promise<string | void>

export type Options = {
  resolver?: (base: string, target: string) => Promise<string | void>,
  loader?: Loader,
  parserOptions?: ParserOptions
}",
      "file": "packages/types/index.d.ts",
    },
    "target": Object {
      "changed": Array [
        Object {
          "end": 2,
          "start": 2,
        },
        Object {
          "end": 14,
          "start": 12,
        },
        Object {
          "end": 19,
          "start": 17,
        },
      ],
      "content": "import { ParserOptions } from '@babel/parser';
import { SourceLocation } from '@babel/types';

export type Module = string;
export type Member = string;
export type Members = Member[] | '*';
export type MemberRef = {
  name: Member,
  alias: Member,
};
export type Import =  MemberRef & {
  source: Module
};
export type Exports = {
  extends?: Module[],
  members: Array<MemberRef & {
    loc: SourceLocation | null
  }>
};
export type MemberRelation =  { [name: string]: Member[] };

export type Entry = {
  source: Module, 
  name: Member
};
export type AffectedMap = Map<Member, Entry[]>;
export type DependencyMap = Map<Module, AffectedMap>;

export type Loader = (path: string) => Promise<string | void>

export type Options = {
  resolver?: (base: string, target: string) => Promise<string | void>,
  loader?: Loader,
  parserOptions?: ParserOptions
}",
      "file": "packages/types/index.d.ts",
    },
  },
]
`;
